drivers: Fix ESP32 I2C unwilled stop and repeated start

znfgnu opened the pull request:
Fixes zephyrproject-rtos/zephyr#40483

- Fixes STOP signal generation condition:
  - at the end of i2c_msg transmission **only** if `I2C_MSG_STOP` flag is set
- Adds new field to driver's data that indicates if STOP condition was generated at the end of the previous transmission.
- Start condition is generated if `I2C_MSG_RESTART` flag is set of if STOP was generated previously.

sylvioalves commented on drivers/i2c/i2c_esp32.c:
```diff
@@ -471,13 +475,13 @@ static int IRAM_ATTR i2c_esp32_write_msg(const struct device *dev,
 		}
 
-		if (msg->len == 0 || (msg->flags & I2C_MSG_STOP)) {
+		if (msg->len == 0 && (msg->flags & I2C_MSG_STOP)) {
```
Hi @znfgnu. When I refactored this, I think I missed removing this check: `msg->len == 0`.  
So this line would be:  
 `if ((msg->flags & I2C_MSG_STOP)) {`  

Would yo be able to re-test only with this change?

znfgnu replied:
Hi @sylvioalves ! I will check this on Monday or Tuesday and let you know.

I think that the change you proposed would work for i2c msgs shorter than `SOC_I2C_FIFO_LEN` bytes.
I don't think it would be correct for i2c messages longer than `SOC_I2C_FIFO_LEN`.
Without checking `msgs->len == 0` STOP would be generated after each `SOC_I2C_FIFO_LEN` bytes.
Please, correct me if I'm wrong.

IMO condition `msg->len == 0 && (msg->flags & I2C_MSG_STOP)` will let STOP to be generated after last transmission (when `msg->len == 0`) for given i2c_msg only when `I2C_MSG_STOP` flag is set, which is expected behavior.

sylvioalves replied:
You are correct. `msg->len` needs to be checked with AND in there. I didn't quite get the reason for having the `data->stop_generated`. What's the scenario for that?

znfgnu replied:
SSD1306 driver (which I used to have problem with) uses `i2c_burst_write` (`drivers/i2c.h`) function to send data to the display. This routine writes `num_bytes` bytes from buffer `*buf` to register `start_addr` of  device with address `dev_addr`. The function implements data transfer by sending two `i2c_msg`s:
```
	msg[0].buf = &start_addr;
	msg[0].len = 1U;
	msg[0].flags = I2C_MSG_WRITE;

	msg[1].buf = (uint8_t *)buf;
	msg[1].len = num_bytes;
	msg[1].flags = I2C_MSG_WRITE | I2C_MSG_STOP;

	return i2c_transfer(dev, msg, 2, dev_addr);
```
Before introducing this flag, esp32 i2c driver used to send repeated start and device address after the first message, which was unintended - `I2C_MSG_RESTART` flag for the second message was not set. In line 461 in my commit I added condition that allows start condition as well as device address to be sent - if `I2C_MSG_RESTART` is set or right after STOP condition was generated.

sylvioalves replied:
I retest the same sample and only changed the `msg->len` check  from OR to AND. It seems it is enough to fix the issue, isn't it? Can you paste the buffer that requires `data->stop_generated` to be added?

Only changing the `msg->len` to AND got this output from I2C: 

```c
I2C data received: 00 AE
I2C data received: 00 D5 80 D9 22 DB 20
I2C data received: 00 40 D3 00 DA 12 A8 3F
I2C data received: 00 A1 C8
I2C data received: 00 8D 14 33
I2C data received: 00 A4 A6
I2C data received: 00 81 80
I2C data received: 00 AF
I2C data received: 00 20 00 21 00 7F 22 00 04
```

znfgnu replied:
It isn't - without check from line 461 the driver sends repeated START with address before sending `msg[1]`'s payload. I'll send you recorded transmission later today (once I go back home).

I suspect that your logging device ignores repeated START and treats two consecutive `i2c_msg`s as one transmission despite the presence of repeated START - maybe because of lack of STOP condition. The display didn't worked in this case.

sylvioalves replied:
@znfgnu thanks, no need to record I transmission.

znfgnu replied:
Despite no need for recording I'll leave it here for other people involved in this MR - it should make my changes more clear.

I checked the driver using `nrf52840dk_nrf52840` and `nucleo_h743zi` boards and transmission looked like the following (display working):
```
Start, 78 [ 3C | WR ], 00, AE, Stop
Start, 78 [ 3C | WR ], 00, D5, 80, D9, 22, DB, 20, Stop
Start, 78 [ 3C | WR ], 00, 40, D3, 00, DA, 12, A8, 3F, Stop
Start, 78 [ 3C | WR ], 00, A1, C8, Stop
Start, 78 [ 3C | WR ], 00, 8D, 14, 33, Stop
Start, 78 [ 3C | WR ], 00, A4, A6, Stop
Start, 78 [ 3C | WR ], 00, 81, 80, Stop
Start, 78 [ 3C | WR ], 00, AF, Stop
Start, 78 [ 3C | WR ], 00, 20, 00, 21, 00, 7F, 22, 00, 04, Stop
```

esp32 without `stop_generated` flag (display not working):
```
Start, 78 [ 3C | WR ], 00, 
Restart, 78 [ 3C | WR ], AE, Stop
Start, 78 [ 3C | WR ], 00, 
Restart, 78 [ 3C | WR ], D5, 80, D9, 22, DB, 20, Stop
Start, 78 [ 3C | WR ], 00, 
Restart, 78 [ 3C | WR ], 40, D3, 00, DA, 12, A8, 3F, Stop
Start, 78 [ 3C | WR ], 00, 
Restart, 78 [ 3C | WR ], A1, C8, Stop
Start, 78 [ 3C | WR ], 00, 
Restart, 78 [ 3C | WR ], 8D, 14, 33, Stop
Start, 78 [ 3C | WR ], 00, 
Restart, 78 [ 3C | WR ], A4, A6, Stop
Start, 78 [ 3C | WR ], 00, 
Restart, 78 [ 3C | WR ], 81, FF, Stop
Start, 78 [ 3C | WR ], 00, 
Restart, 78 [ 3C | WR ], AF, Stop
Start, 78 [ 3C | WR ], 00, 
Restart, 78 [ 3C | WR ], 20, 00, 21, 00, 7F, 22, 00, 04, Stop
```

esp32 with `stop_generated` flag (display working):
```
Start, 78 [ 3C | WR ], 00, AE, Stop
Start, 78 [ 3C | WR ], 00, D5, 80, D9, 22, DB, 20, Stop
Start, 78 [ 3C | WR ], 00, 40, D3, 00, DA, 12, A8, 3F, Stop
Start, 78 [ 3C | WR ], 00, A1, C8, Stop
Start, 78 [ 3C | WR ], 00, 8D, 14, 33, Stop
Start, 78 [ 3C | WR ], 00, A4, A6, Stop
Start, 78 [ 3C | WR ], 00, 81, 80, Stop
Start, 78 [ 3C | WR ], 00, AF, Stop
Start, 78 [ 3C | WR ], 00, 20, 00, 21, 00, 7F, 22, 00, 04, Stop
```

sylvioalves commented:
@znfgnu Although the proposed PR works, I think we could have this fix by handling it in its messages flags instead of creating a new variable (very similar to other I2C drivers). Could you PTAL in #40651? I would copy the diff in here but it it will mess up). If you approve, feel free to change in here.

znfgnu committed:
drivers: i2c: fix esp32 i2c driver unwilled STOP and repeated START

- added messages check and fix before transfer
- fixed START and STOP signal generation conditions
- fixed semaphore release for empty messages

Signed-off-by: Konrad Sikorski <znfgnu@gmail.com>

znfgnu commented:
@sylvioalves I fixed the driver according to changes you proposed. The difference is that messages validation is done before the transfer. By the way, I moved `k_sem_take` after emptiness check - for empty transmissions the semaphore used to be taken but not released.

sylvioalves commented:
@ulipe @glaubermaroto PTAL

sylvioalves commented:
@znfgnu Please, add the commit signature.

znfgnu commented:
> @znfgnu Please, add the commit signature.

Done by force-pushing amended commit. Let me know if there is any other, more human way to do it.
